torch/fx/passes/infra/partitioner.py

```diff
@@ -69,38 +111,33 @@ def maybe_merge_partition(self_id: int, other_id: int):
-            # has been visited. We are NEVER going to iterate on nodes inside
-            # the set.
-            visited: Set[Node] = set()
-
             def dfs_iter_find_cycle(root_node):
                 stack : Deque[Node] = deque()
                 stack.append(root_node)
-
-                while stack:
-                    node = stack.pop()
-
-                    if node in visited:
-                        continue
-                    if node in merged_nodes:
-                        return True  # found cycle, return
-
-                    # branching on hitting partition or not
-                    if node in assignment:
-                        # Since partition is not merged in the graph yet, when we
-                        # hit a node in a partition through DFS, we need to
-                        # traverse all nodes in the partition to properly reflect
-                        # dependencies after the fusion
-                        for p_node in partitions_by_id[assignment[node]].nodes:
-                            for user_node in p_node.users:
-                                if user_node not in partitions_by_id[assignment[node]].nodes:
-                                    stack.append(user_node)
-                    else:
-                        for user_node in node.users:
-                            stack.append(user_node)
-
-                    visited.add(node)
+                visited_partition_ids = set()
+
+                for path_node in dfs_paths_from_node[root_node]:
+                    # If any of the nodes in the dfs path of this node are in the merged_nodes
+                    # list then there is a cycle in the graph.
+                    if path_node in merged_nodes:
+                        return True
+
+                    # If any of the nodes in the dfs path of this node are in the assignment
+                    # map then we have to make sure that the partitions that these nodes belong
+                    # to do not form a cycle with the current partitions being merged. This means
+                    # iterating through all the nodes in all the parititons that are traversed in
+                    # the dfs path and checking if they are in the merged_nodes list.
+                    if path_node in assignment:
+                        partition_id = assignment[path_node]
+                        # If the partition id has already been visited then we know that it doesn't
+                        # form a cycle with the current partitions being merged.
+                        if partition_id in visited_partition_ids:
+                            continue
+                        p_map = partition_map[partition_id]
+                        if self_id in p_map or other_id in p_map:
+                            return True
+
+                        visited_partition_ids.add(partition_id)
 
                 return False
 
```

Comment:
I think the line 108~113 can be rewritten to only call `dfs_iter_find_cycle` once. 
- gather all user_nodes of "merged_nodes"
- pass this list into `dfs_iter_find_cycle`

```
            for node in merged_nodes:
                for user_node in node.users:
                    if user_node not in merged_nodes:
                        all_user_nodes.appen(user_node)   
                if dfs_iter_find_cycle(all_user_node):
                        # return false indicating cyclic dependency found and
                        # merge is aborted
                        return False
```

Reply:
That's a good idea. Made the change.